src/main/java/com/google/api/codegen/discovery/transformer/csharp/CSharpSampleTypeNameConverter.java

```diff
@@ -71,19 +75,47 @@ public TypeName getServiceTypeName(String apiTypeName) {
 
   @Override
   public TypeName getRequestTypeName(String apiTypeName, TypeInfo typeInfo) {
-    return getTypeName(typeInfo);
+    List<String> copy = new ArrayList<>(methodNameComponents);
+    String requestTypeName = copy.remove(copy.size() - 1) + "Request";
+    String path = "";
+    for (String s : copy) {
+      path += s + "Resource.";
+    }
+    return new TypeName(path + requestTypeName);
```

Comment:
It's not really a change, I just moved the logic that decides these names from the `CSharpTypeNameGenerator`. The behavior is the exact same. I've been whittling down the language specific type name generators bit by bit and moving their logic to the converters.